### PR TITLE
Implement Dirichlet and Beta distributions

### PR DESCRIPTION
**Pull request by fritzo**
This Implements `Dirichlet` and `Beta` distributions. The samplers use`Gamma.sample()` and the reparameterized gradients are computed with numerical approximations to the `cdf`. The computed gradients are accurate to within 1% relative error for a wide range of parameters.

For details of the derivation, see this [Jupyter Notebook](https://github.com/fritzo/notebooks/blob/master/beta-reparameterized.ipynb). Math has been reviewed by @martinjankowiak

### CODE REVIEW

File: torch/distributions.py [hunk 1]
```diff
@@ -262,3 +262,107 @@ def log_prob(self, value):
...
+
+    Args:
+        alpha (Tensor or Variable): concentration parameter of the distribution
+    """
+    has_rsample = True
+
+    def __init__(self, alpha):
+        self.alpha = alpha
+
+    def _sample(self, alpha):
```
Comment (apaszke) [inline torch/distributions.py]: Is this needed? It seems to me that `sample_n` could just call `sample` if you moved this code there

File: torch/distributions.py [hunk 2]
```diff
@@ -262,3 +262,107 @@ def log_prob(self, value):
...
+        if alpha_num and beta_num:
+            alpha_beta = torch.Tensor([alpha, beta])
+        else:
+            if alpha_num and not beta_num:
+                alpha = beta.new(beta.size()).fill_(alpha)
+            elif not alpha_num and beta_num:
+                beta = alpha.new(alpha.size()).fill_(beta)
+            elif alpha.size() != beta.size():
+                raise ValueError('Expected alpha.size() == beta.size(), actual {} vs {}'.format(
+                    alpha.size(), beta.size()))
```
Comment (apaszke) [inline torch/distributions.py]: This seems to be a recurring pattern so it would be nice to have a helper for this. Neeraj had a neat way of dealing with it (upcasting everything to tensors, and then a tiny helper for broadcasting them together)

**Review by apaszke — COMMENTED** [review]
LGTM. Some minor comments

**Comment by fritzo** [comment]
@martinjankowiak could you please review the math?

**Comment by martinjankowiak** [comment]
@fritzo yup just give me a bit of time

File: torch/distributions.py (hunk 1 above)
Comment (fritzo) [reply torch/distributions.py]: This helper is needed until #4100 is merged. Currently `.sample_n()` would not be able to change the shape of `self.alpha` for use inside `.sample()`. Once `.sample()` can accept an `n` or `sample_shape` parameter, we can combine all these methods into a single `.sample(sample_shape)` method.

**Comment by martinjankowiak** [comment]
ok. so we've both gone over the math carefully offline. LGTM!

File: torch/distributions/beta.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        alpha (Tensor or Variable): concentration parameter of the distribution
+    """
+    has_rsample = True
+
+    def __init__(self, alpha, beta):
+        if isinstance(alpha, Number) and isinstance(beta, Number):
+            alpha_beta = torch.Tensor([alpha, beta])
+        else:
+            alpha, beta = broadcast_all(alpha, beta)
+            alpha_beta = torch.stack([alpha, beta], -1)
```
Comment (apaszke) [inline torch/distributions/beta.py]: Is this special handling for scalars expected? I thought we decided to just treat them as 1d arrays everywhere for consistency. Not a problem for merging, just wanted to make sure it's not like this by mistake.

**Review by apaszke — COMMENTED** [review]
LGTM, only some minor things, but should be ready to merge. @gchanan should we wait with this until ATen apply is merged, or can we port it later?

File: test/test_distributions.py [hunk 4]
```diff
@@ -262,6 +271,96 @@ def test_gamma_sample_grad(self):
                                        'rel error {}'.format(rel_error),
                                        'max error {}'.format(rel_error.max())]))
 
+    def test_dirichlet_shape(self):
+        alpha = Variable(torch.exp(torch.randn(2, 3)), requires_grad=True)
+        alpha_1d = Variable(torch.exp(torch.randn(4)), requires_grad=True)
+        self.assertEqual(Dirichlet(alpha).sample().size(), (2, 3))
+        self.assertEqual(Dirichlet(alpha).sample_n(5).size(), (5, 2, 3))
+        self.assertEqual(Dirichlet(alpha_1d).sample().size(), (4,))
+        self.assertEqual(Dirichlet(alpha_1d).sample_n(1).size(), (1, 4))
```
[the remark this answers is not in the document]
Comment (fritzo) [reply test/test_distributions.py]: I agree. I'll send a separate PR updating all the tests and will update this PR.

File: torch/distributions/beta.py (hunk 3 above)
Comment (fritzo) [reply torch/distributions/beta.py]: This is intended to match the behavior of other distributions. This special handling will no longer be needed once we have scalar support; until then you can get scalar-like shapes by passing `float`s as parameters.

File: test/test_distributions.py [hunk 5]
```diff
@@ -1,10 +1,13 @@
...
-import torch
 import unittest
+import warnings
 from itertools import product
+
+import torch
+from common import TestCase, run_tests
 from torch.autograd import Variable, gradcheck
-from torch.distributions import Bernoulli, Categorical, Normal, Gamma, Distribution
+from torch.distributions import (Bernoulli, Beta, Categorical, Dirichlet,
```
[the remark this answers is not in the document]
Comment (fritzo) [reply test/test_distributions.py]: I'll switch to that if you prefer.

[the remark this answers is not in the document]
Comment (fritzo) [reply test/test_distributions.py]: Egads, that leads to lots of `pyflakes` warnings. Would you mind if we keep the exhaustive list?

[the remark this answers is not in the document]
Comment (fritzo) [reply test/test_distributions.py]: We never use `import *` in Pyro and I avoid it in my own code since it's so dangerous. E.g. it's unclear whether a symbol from `import *` shadows a previous import.

[the remark this answers is not in the document]
Comment (apaszke) [reply test/test_distributions.py]: That is true, although we're using them in a few places and didn't have problems so far. It was just a suggestion, I'm ok with listing them explicitly

[the remark this answers is not in the document]
Comment (fritzo) [reply test/test_distributions.py]: Ok thanks.

**Comment by fritzo** [comment]
🎉 Thanks, @apaszke !
